### PR TITLE
Allow path syntax for super constructor calls

### LINKED ISSUE #9495: Allow the member notation (M.C) when specifying (non-constant) parent constructors
Would it be possible to allow the member notation (M.C) when specifying (non-constant) parent constructors ?

This would be useful in the following scenario:
```
// file1.sol

contract C {
  int private x;
  constructor (int p) public { x = p; }
}

// file2.sol

import "file1.sol" as M;

contract D is M.C {
  constructor (int p) M.C(p) public {} // M.C currently not allowed in this context
}
```

For now, I rely on the following workaround, but it's not ideal:
```
import "file1.sol" as M;
import { C as M_C } from "file1.sol";

contract D is M.C {
  constructor (int p) M_C(p) public {}
}
```

### PR DESCRIPTION
**Pull request by Marenz**
fixes #9495

### CODE REVIEW

File: libsolidity/analysis/DeclarationTypeChecker.cpp [hunk 1]
```diff
@@ -276,6 +278,14 @@ void DeclarationTypeChecker::endVisit(VariableDeclaration const& _variable)
 	if (_variable.annotation().type)
 		return;
 
+	if (auto typeName = dynamic_cast<UserDefinedTypeName const*>(&_variable.typeName()))
+	   if (dynamic_cast<ModifierDefinition const*>(typeName->annotation().referencedDeclaration))
```
Comment (stackenbotten) [inline libsolidity/analysis/DeclarationTypeChecker.cpp]: Coding style error

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `a0f868b64c17b7e68febfca9e2a70339f67818c8`:
```
Coding style error:
 libsolidity/analysis/DeclarationTypeChecker.cpp:282: if (dynamic_cast<ModifierDefinition const*>(typeName->annotation().referencedDeclaration))

```
Please check that your changes are working as intended.

File: scripts/isolate_tests.py [hunk 2]
```diff
@@ -43,7 +43,7 @@ def extract_docs_cases(path):
     tests = []
 
     # Collect all snippets of indented blocks
-    for l in open(path, mode='r', encoding='utf8').read().splitlines():
+    for l in open(path, mode='r', errors='ignore', encoding='utf8').read().splitlines():
```
Comment (chriseth) [inline scripts/isolate_tests.py]: Should this be its own PR?

File: libsolidity/analysis/TypeChecker.cpp [hunk 3]
```diff
@@ -3046,6 +3046,12 @@ bool TypeChecker::visit(Identifier const& _identifier)
 	return false;
 }
 
+void TypeChecker::endVisit(UserDefinedTypeName const& _userDefinedTypeName)
+{
+	if (!_userDefinedTypeName.annotation().type)
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp:3051]: When would this be the case?

File: libsolidity/analysis/ImmutableValidator.cpp [hunk 4]
```diff
@@ -120,6 +120,12 @@ bool ImmutableValidator::visit(WhileStatement const& _whileStatement)
 	return false;
 }
 
+void ImmutableValidator::endVisit(UserDefinedTypeName const& _userDefinedTypeName)
+{
+	if (auto const callableDef = dynamic_cast<CallableDeclaration const*>(_userDefinedTypeName.annotation().referencedDeclaration))
+		visitCallableIfNew(callableDef->resolveVirtual(m_currentContract));
```
Comment (chriseth) [inline libsolidity/analysis/ImmutableValidator.cpp:126]: I'm not sure if we should do virtual lookup here.

File: libsolidity/analysis/DeclarationTypeChecker.cpp (hunk 1 above)
Comment (chriseth) [inline libsolidity/analysis/DeclarationTypeChecker.cpp]: I think we should rather do a positive list instead of a negative list here.

Comment (chriseth) [reply libsolidity/analysis/DeclarationTypeChecker.cpp]: We might also want to move this down to where `typeName` is accessed.

File: libsolidity/analysis/DeclarationTypeChecker.cpp [hunk 5]
```diff
@@ -146,7 +146,9 @@ void DeclarationTypeChecker::endVisit(UserDefinedTypeName const& _typeName)
 		_typeName.annotation().type = TypeProvider::enumType(*enumDef);
 	else if (ContractDefinition const* contract = dynamic_cast<ContractDefinition const*>(declaration))
 		_typeName.annotation().type = TypeProvider::contract(*contract);
-	else
+	// Can't assign the type for modifiers at this stage as it needs info from
```
Comment (chriseth) [inline libsolidity/analysis/DeclarationTypeChecker.cpp:149]: I'm actually not sure we should even get here. This is really the context of a type. Maybe we should split this into UserDefinedTypeName and `Path` (which is just a list of identifiers separated by a dot instead of something really signifying a type)

File: libsolidity/analysis/ImmutableValidator.cpp (hunk 4 above)
Comment (Marenz) [reply libsolidity/analysis/ImmutableValidator.cpp:126]: Well, this is exactly the same thing as we're doing when visiting the identifiers which is what was previously used for the super constructor calls. If we shouldn't do this here then we shouldn't do it there either.

File: libsolidity/analysis/DeclarationTypeChecker.cpp (hunk 1 above)
Comment (Marenz) [reply libsolidity/analysis/DeclarationTypeChecker.cpp]: If we move this down we get assert failures because the location info is not available but accessed a few lines down.
With list you mean we should name the things a variable can be instead of that it can't be a modifier?

File: libsolidity/analysis/DeclarationTypeChecker.cpp (hunk 5 above)
Comment (Marenz) [reply libsolidity/analysis/DeclarationTypeChecker.cpp:149]: Well, the UserDefinedTypeName is a parser construct which can't differentiate between those too..

File: libsolidity/analysis/TypeChecker.cpp (hunk 3 above)
Comment (Marenz) [reply libsolidity/analysis/TypeChecker.cpp:3051]: Usually the `DeclarationTypeChecker::visit(UserDefinedTypeName..)` would assign the type, but we can't do that for modifiers because they require information only available in this stage so for the modifiers we have to do it here.

File: scripts/isolate_tests.py (hunk 2 above)
Comment (Marenz) [reply scripts/isolate_tests.py]: I can pull it out. I just stumbled over that while testing and quickly fixed it

File: libsolidity/analysis/ImmutableValidator.cpp (hunk 4 above)
Comment (chriseth) [reply libsolidity/analysis/ImmutableValidator.cpp:126]: It is different because we do not do virtual lookup for functions that use path notation either. The idea behind is that this is always qualified access: `a.x` means "the x defined in a" while `x` means "whatever x currently is overwritten by"

Comment (Marenz) [reply libsolidity/analysis/ImmutableValidator.cpp:126]: Mind you, this is also for modifiers, which can have virtual behavior (if I recall correctly)

Comment (Marenz) [reply libsolidity/analysis/ImmutableValidator.cpp:126]: In this case though, the AST type that was previously `Identifier` changed to `UserDefinedTypeName`, so even though it might just be a single identifier, it still is of the UserDefinedTypeName type. 

We can add a check to see if it is actually a path or just a single word and behave differently based on that...

File: libsolidity/analysis/DeclarationTypeChecker.cpp [hunk 6]
```diff
@@ -276,6 +278,14 @@ void DeclarationTypeChecker::endVisit(VariableDeclaration const& _variable)
 	if (_variable.annotation().type)
 		return;
 
+	if (auto typeName = dynamic_cast<UserDefinedTypeName const*>(&_variable.typeName()))
+	   if (dynamic_cast<ModifierDefinition const*>(typeName->annotation().referencedDeclaration) != nullptr)
```
Comment (stackenbotten) [inline libsolidity/analysis/DeclarationTypeChecker.cpp]: Coding style error

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `3518432e4bbe7fea52497f9eef9a77d8f3cbefe6`:
```
Coding style error:
 libsolidity/analysis/DeclarationTypeChecker.cpp:282: if (dynamic_cast<ModifierDefinition const*>(typeName->annotation().referencedDeclaration) != nullptr)

```
Please check that your changes are working as intended.

File: libsolidity/analysis/DeclarationTypeChecker.cpp (hunk 5 above)
Comment (chriseth) [reply libsolidity/analysis/DeclarationTypeChecker.cpp:149]: Actually I think the reason we got into this problem is that modifiers are much closer to the expression context than to the type context. So this should maybe even be a MemberAccess instead.

**Comment by chriseth** [comment]
This seems to be much more complicated than initially anticipated. I think we should use UserDefinedTypeName really just for type names and thus ModifierInvocation should use an Expression instead of a UserDefinedTypeName. Maybe we can find a middle ground where we do not allow the full power of a generic expression - I think there are other places where we use UserDefinedTypeName that are actually not really type names.

**Comment by chriseth** [comment]
Maybe the solution is to use something like `IdentifierPath` inside UserDefinedTypeName (which inherits from TypeName and has the corresponding annotation), and use IdentifierPath directly in all the other AST nodes that are not really types like InheritanceSpecifier, UsingForDirective and OverrideSpecifier.

**Comment by Marenz** [comment]
replaced by #9753
